Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs

```diff
@@ -1,347 +1,187 @@
+            StartLoop();
+        }
 
-			// check the size of pending in memory
+        public void AddToPending(Block block)
+        {
+            ChainedBlock chainedBlock = this.Chain.GetBlock(block.GetHash());
+            if (chainedBlock == null)
+                return; // reorg
 
-			// add to pending blocks
-			if (this.StoredBlock.Height < chainedBlock.Height)
-				this.PendingStorage.TryAdd(chainedBlock.HashBlock, new BlockPair {Block = block, ChainedBlock = chainedBlock});
-		}
+            // check the size of pending in memory
 
-		public Task Flush()
-		{
-			return this.DownloadAndStoreBlocks(CancellationToken.None, true);
-		}
+            // add to pending blocks
+            if (this.StoredBlock.Height < chainedBlock.Height)
+                this.PendingStorage.TryAdd(chainedBlock.HashBlock, new BlockPair(block, chainedBlock));
+        }
 
+        public Task Flush()
+        {
+            return DownloadAndStoreBlocks(CancellationToken.None, true);
+        }
 
-		public void StartLoop()
-		{
-			// A loop that writes pending blocks to store 
-			// or downloads missing blocks then writing to store
-			this.asyncLoopFactory.Run("BlockStoreLoop.DownloadBlocks", async token =>
-			{
-				await this.DownloadAndStoreBlocks(this.nodeLifetime.ApplicationStopping);
-			},
-			this.nodeLifetime.ApplicationStopping,
-			repeatEvery: TimeSpans.Second,
-			startAfter: TimeSpans.FiveSeconds);
-		}
+        /// <summary>
+        /// A loop that:
+        ///     1: Write pending blocks to store 
+        ///     2: Download missing blocks and write to store
+        /// </summary>
+		internal void StartLoop()
+        {
+            this.asyncLoopFactory.Run("BlockStoreLoop.DownloadBlocks", async token =>
+                {
+                    await DownloadAndStoreBlocks(this.nodeLifetime.ApplicationStopping);
+                },
+                this.nodeLifetime.ApplicationStopping,
+                repeatEvery: TimeSpans.Second,
+                startAfter: TimeSpans.FiveSeconds);
+        }
 
-		public async Task DownloadAndStoreBlocks(CancellationToken token, bool disposemode = false)
-		{
+        private async Task DownloadAndStoreBlocks(CancellationToken cancellationToken, bool disposeMode = false)
+        {
             // TODO: add support to BlockStoreLoop to unset LazyLoadingOn when not in IBD
             // When in IBD we may need many reads for the block key without fetching the block
             // So the repo starts with LazyLoadingOn = true, however when not anymore in IBD 
             // a read is normally done when a peer is asking for the entire block (not just the key) 
             // then if LazyLoadingOn = false the read will be faster on the entire block            
 
-            while (!token.IsCancellationRequested)
-			{
-				if (this.StoredBlock.Height >= this.ChainState.HighestValidatedPoW?.Height)
-					break;
+            var steps = new BlockStoreLoopStepChain(disposeMode);
```

Comment:
Yes, the step chain will be created on every loop. I will move this to the Initialize() method.